freeipa/src/test/java/com/sequenceiq/freeipa/service/BootstrapServiceTest.java

```diff
@@ -0,0 +1,288 @@
+                hasProperty("hostGroup", is("TADA"))
+                        )));
+        assertThat(allNodes, hasItem(
+                allOf(hasProperty("instanceId", is(INSTANCE_WO_FQDN)),
+                hasProperty("hostname", is(HOSTNAME + '1')),
+                hasProperty("domain", is(DOMAIN)),
+                hasProperty("instanceType", is("GW")),
+                hasProperty("hostGroup", is("TADA"))
+                        )));
+        assertThat(allNodes, hasItem(
+                allOf(hasProperty("instanceId", is(INSTANCE_WRONG_DOMAIN)),
+                hasProperty("hostname", is(HOSTNAME + '1')),
+                hasProperty("domain", is(DOMAIN)),
+                hasProperty("instanceType", is("GW")),
+                hasProperty("hostGroup", is("TADA"))
+                        )));
+        BootstrapParams bootstrapParams = bootstrapParamsCaptor.getValue();
+        assertTrue(bootstrapParams.isSaltBootstrapFpSupported());
+        assertTrue(bootstrapParams.isRestartNeededFlagSupported());
+        assertEquals(image.getOs(), bootstrapParams.getOs());
+        assertEquals(stack.getCloudPlatform(), bootstrapParams.getCloud());
+        StackBasedExitCriteriaModel exitCriteriaModel = (StackBasedExitCriteriaModel) exitCriteriaModelCaptor.getValue();
+        assertEquals(STACK_ID, exitCriteriaModel.getStackId().get());
+    }
+
+    @Test
+    public void testIOExceptionConverted() throws CloudbreakOrchestratorException, IOException {
+        when(instanceMetaDataRepository.findNotTerminatedForStack(STACK_ID)).thenReturn(Set.of(
+                createInstance(INSTANCE_WITH_FQDN, "instance1" + DOMAIN),
+                createInstance(INSTANCE_WO_FQDN, null),
+                createInstance(INSTANCE_WRONG_DOMAIN, "instance.wrong.domain")));
+        Stack stack = new Stack();
+        stack.setCloudPlatform("cloud");
+        when(stackRepository.findById(STACK_ID)).thenReturn(Optional.of(stack));
+        FreeIpa freeIpa = new FreeIpa();
+        freeIpa.setDomain(DOMAIN);
+        freeIpa.setHostname(HOSTNAME);
+        when(freeIpaService.findByStack(stack)).thenReturn(freeIpa);
+        List<GatewayConfig> gatewayConfigs = List.of();
+        when(gatewayConfigService.getGatewayConfigs(eq(stack), anySet())).thenReturn(gatewayConfigs);
+        byte[] bytes = {};
+        when(compressUtil.generateCompressedOutputFromFolders("salt-common", "freeipa-salt")).thenReturn(bytes);
+        ImageEntity image = new ImageEntity();
+        image.setOs("ZOS");
+        when(imageService.getByStack(stack)).thenReturn(image);
+        when(hostDiscoveryService.generateHostname(anyString(), any(), anyLong(), anyBoolean())).thenCallRealMethod();
+        doThrow(new CloudbreakOrchestratorCancelledException("wow")).when(hostOrchestrator).bootstrapNewNodes(any(), anySet(), anySet(), any(), any(), any());
+
+        assertThrows(CloudbreakOrchestratorException.class, () -> underTest.bootstrap(STACK_ID));
```

Comment:
yeah, but I only test exception message where is logic related to it. If you insist I can add it, but I don't think it's necessary here

Reply:
The logic there is thin, true, the main purpose would only to check that the original exception and its message is not hidden. However, this could go into spotBugs. I once checked that 2 years ago, back then this was on a PR in a related repo, it might make sense to revisit it => I added a jira for that: https://jira.cloudera.com/browse/CB-13613